File: src/main/java/net/benwoodworth/fastcraft/core/api/dependencies/inventory/FcItem.java
package net.benwoodworth.fastcraft.core.api.dependencies.inventory;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * An immutable Minecraft item.
 *
 * @param <TBase> the native item type
 */
public abstract class FcItem<TBase> {

    @NotNull
    protected final TBase baseItem;

    /**
     * Create a new immutable FcItem.
     *
     * @param baseItem the base item
     */
    protected FcItem(@NotNull TBase baseItem) {
        this.baseItem = baseItem;
    }

    /**
     * Construct an immutable FcItem from another FcItem.
     *
     * @param item the FcItem this will be based off
     */
    public FcItem(@NotNull FcItem<TBase> item) {
        this(item.cloneBaseItem());
    }

    /**
     * Clone the base item being adapted.
     *
     * @return a clone of the base item
     */
    @NotNull
    public abstract TBase cloneBaseItem();

    /**
     * Get the Minecraft item type ID.
     *
     * @return the item type ID
     */
    @NotNull
    public abstract String getTypeId();

    /**
     * Get the amount of items in this stack.
     *
     * @return the amount
     */
    public abstract int getAmount();

    /**
     * Get the display name of this item.
     *
     * @return the item display name, or {@code null} if there is none
     */
    @Nullable
    public abstract String getDisplayName();

    /**
     * Get a copy of this item's lore.
     *
     * @return a copy of the lore
     */
    @NotNull
    public abstract List<String> getLore();

    /**
     * Get the maximum size of this stack.
     *
     * @return the maximum stack size
     */
    public abstract int getMaxStackSize();

    /**
     * Whether this item, as an ingredient, matches any item of the same type, regardless of data.
     *
     * @return {@code true} if the item has wild card data
     */
    public abstract boolean hasWildCardData();

    /**
     * Compare equality of this base item to another, ignoring amount.
     *
     * @param item the {@link FcItem} to compare to
     * @return {@code true} if the items are similar
     */
    public abstract boolean isSimilar(@NotNull FcItem<TBase> item);

    /**
     * Checks if this item can be used as an ingredient in a recipe.
     * Ingredient can have wildcard data.
     *
     * @param ingredient the ingredient of a recipe
     * @return {@code true} if this item can be used as the ingredient
     */
    public abstract boolean matchesIngredient(@NotNull FcItem<TBase> ingredient);

    /**
     * A hash code for this item.
     *
     * @return the item's hash code
     */
    @Override
    public abstract int hashCode();

    /**
     * Compare equality to an object that isn't an FcItem<T>.
     *
     * @return {@code false}
     */
    @Override
    public final boolean equals(@Nullable Object other) {
        return false;
    }

    /**
     * Checks if the items are similar, and have the same amount.
     *
     * @return {@code true} if the items are equal
     */
    public final boolean equals(@NotNull FcItem<TBase> other) {
        return getAmount() == other.getAmount() && isSimilar(other);
    }
}

File: src/main/kotlin/net/benwoodworth/fastcraft/core/api/dependencies/inventory/FcItem.kt
package net.benwoodworth.fastcraft.core.api.dependencies.inventory

/**
 * An immutable Minecraft item.
 *
 * @param TBase the native item type
 */
abstract class FcItem<TBase>(
        /** The base item */
        open protected val baseItem: TBase
) {

    /** The Minecraft item type ID. */
    abstract var typeId: String

    /** The amount of items in this stack. */
    abstract var amount: Int

    /** The display name of this item. */
    abstract var displayName: String?

    /** The maximum size of this stack. */
    abstract val maxStackSize: Int

    /**
     * Construct an immutable FcItem from another FcItem.
     *
     * @param item the FcItem this will be based off
     */
    constructor(item: FcItem<TBase>) : this(item.cloneBaseItem())

    /**
     * Clone the base item being adapted.
     *
     * @return a clone of the base item
     */
    abstract fun cloneBaseItem(): TBase

    /**
     * Get the item's lore.
     *
     * @return the item's lore
     */
    abstract fun getLore(): List<String?>

    /**
     * Set the item's lore.
     *
     * @param lore the item's lore
     */
    abstract fun setLore(lore: List<String?>)

    /**
     * Add an enchantment to the item.
     *
     * @param enchantmentId the Minecraft enchantment ID
     * @param level the enchantment level
     */
    abstract fun addEnchantment(enchantmentId: String, level: Int)

    /**
     * Whether this item, as an ingredient, matches any
     * item of the same type, regardless of data.
     */
    abstract val hasWildCardData: Boolean

    /**
     * Compare equality of this base item to another, ignoring amount.
     *
     * @param item the [FcItem] to compare to
     * @return `true` if the items are similar
     */
    abstract fun isSimilar(item: FcItem<TBase>): Boolean

    /**
     * Checks if this item can be used as an ingredient in a recipe.
     * Ingredient can have wildcard data.
     *
     * @param ingredient the ingredient of a recipe
     * @return `true` if this item can be used as the ingredient
     */
    abstract fun matchesIngredient(ingredient: FcItem<TBase>): Boolean

    /**
     * A hash code for this item.
     *
     * @return the item's hash code
     */
    abstract override fun hashCode(): Int

    /**
     * Compare equality to an object that isn't an FcItem<T>.
     *
     * @return `false`
     */
    final override fun equals(other: Any?) = false

    /**
     * Checks if the items are similar, and have the same amount.
     *
     * @return `true` if the items are equal
     */
    fun equals(other: FcItem<TBase>): Boolean {
        return amount == other.amount && isSimilar(other)
    }
}
